### ISSUE TITLE
latent worker is rejected if buildbot master is reconfigured during preparing worker

### ISSUE THREAD

**Issue by jonnyyu** [buildbot/buildbot#3438]
## Environment

buildbot v0.9.8  + some latent worker (we have our own implementation, but I think ec2 can repro too)
Repro steps:
1. trigger some jobs to launch latent worker
2. during preparing worker, before worker attached,  do buildbot reconfig
### Expect:
buildbot worker execute the job normally
### Actual:
buildbot job staying at 'preparing worker' phase,  worker's twistd.log shows:

```
2017-07-13 19:38:20-0700 [HangCheckProtocol,client] message from master: attached
2017-07-13 19:38:20-0700 [HangCheckProtocol,client] ReconnectingPBClientFactory.failedToGetPerspective
2017-07-13 19:38:20-0700 [HangCheckProtocol,client] While trying to connect:
	Traceback from remote host -- exceptions.RuntimeError: Worker worker_018 received connection while not trying to substantiate.  Disconnecting.
```

## Analysis:

AbstractLatentWorker.substantiate() method calls _substantiation_notifier.wait() to add a waiter to Notifier object before calling _substantiate() method. _substantiate() method usually would take quite some time (several minutes to 10-20 mins), if reconfigure happens during that time, AbstractLatentWorker.reconfigService always creates new Notifier object for _substantiation_notifier, the previous notifier which has waiters is replaced by the new Notifier object without any waiters. Thus worker gets disconnected when AbstractLatentWorker.attached method checks AbstractLatentWorker._substantiation_notifier has no waiters.

**Comment by tardyp** [comment]
Indeed, I guess this is a bug that will be especially annoying when you have very workers that take long to start.
Would you like to try and submit a fix?

You can look at those integration tests:
https://github.com/buildbot/buildbot/blob/master/master/buildbot/test/integration/test_latent.py#L318
With the help of those, I believe it will be quite simple to reproduce your issue, and then you can try and fix it by keeping the other tested behaviour working, and we can discuss the result in a PR.

**Comment by kpweiler** [comment]
I think I'm running into this same issue, but I'm not getting the same logs. This is maybe not too surprising as this issue is rather old. Is this recent commit related?

https://github.com/buildbot/buildbot/commit/57d494271ef37085345f23b126cbb088b457cd65

Here are the relevant log messages:

#### master:
```
2020-03-04 13:15:30-0600 [-] worker 'latent_9' attaching from IPv4Address(type='TCP', host='172.18.100.55', port=47410)
2020-03-04 13:15:30-0600 [-] configuration update complete
2020-03-04 13:15:30-0600 [Broker,7277,172.18.100.55] Got workerinfo from 'latent_9'
2020-03-04 13:15:30-0600 [-] bot attached
2020-03-04 13:15:30-0600 [Broker,7277,172.18.100.55] Worker latent_9 attached to core-test
2020-03-04 13:15:30-0600 [Broker,7277,172.18.100.55] Worker latent_9 attached to core
2020-03-04 13:15:30-0600 [Broker,7277,172.18.100.55] Worker latent_9 attached to datacube-release
2020-03-04 13:15:30-0600 [Broker,7277,172.18.100.55] Worker latent_9 attached to core-env
2020-03-04 13:15:30-0600 [Broker,7277,172.18.100.55] Worker latent_9 attached to edgarparsers-release
2020-03-04 13:15:30-0600 [Broker,7277,172.18.100.55] Worker latent_9 attached to buildbot-app
2020-03-04 13:15:30-0600 [Broker,7277,172.18.100.55] Worker latent_9 attached to chantey
2020-03-04 13:15:30-0600 [Broker,7277,172.18.100.55] Worker latent_9 attached to lake-server
2020-03-04 13:15:30-0600 [Broker,7277,172.18.100.55] Worker latent_9 attached to yard
2020-03-04 13:15:30-0600 [Broker,7277,172.18.100.55] Worker latent_9 substantiated \o/
2020-03-04 13:15:30-0600 [Broker,7277,172.18.100.55] No substantiation deferred for latent_9
```

#### worker:
```
Agent pid 24
2020-03-04 19:15:29+0000 [-] Loading buildbot.tac...
2020-03-04 19:15:29+0000 [-] Loaded.
2020-03-04 19:15:29+0000 [-] twistd 19.10.0 (/usr/bin/python3 3.6.9) starting up.
2020-03-04 19:15:29+0000 [-] reactor class: twisted.internet.epollreactor.EPollReactor.
2020-03-04 19:15:29+0000 [-] Starting Worker -- version: 2020.02.27
2020-03-04 19:15:29+0000 [-] recording hostname in twistd.hostname
2020-03-04 19:15:29+0000 [-] Locating buildmaster...
2020-03-04 19:15:30+0000 [-] Starting factory <buildbot_worker.pb.BotFactory object at 0x7f2ce9afc518>
2020-03-04 19:15:30+0000 [Broker,client] message from master: attached
2020-03-04 19:15:30+0000 [Broker,client] I have a leftover directory 'twisted' that is not being used by the buildmaster: you can delete it now
2020-03-04 19:15:30+0000 [Broker,client] message from master: attached
2020-03-04 19:15:30+0000 [Broker,client] message from master: attached
2020-03-04 19:15:30+0000 [Broker,client] message from master: attached
2020-03-04 19:15:30+0000 [Broker,client] message from master: attached
2020-03-04 19:15:30+0000 [Broker,client] message from master: attached
2020-03-04 19:15:30+0000 [Broker,client] message from master: attached
2020-03-04 19:15:30+0000 [Broker,client] message from master: attached
2020-03-04 19:15:30+0000 [Broker,client] message from master: attached
2020-03-04 19:15:30+0000 [Broker,client] message from master: attached
2020-03-04 19:15:30+0000 [Broker,client] Connected to buildmaster; worker is ready
2020-03-04 19:15:30+0000 [Broker,client] sending application-level keepalives every 600 seconds
2020-03-04 19:25:30+0000 [-] sending app-level keepalive
2020-03-04 19:25:30+0000 [Broker,client] Master replied to keepalive, everything's fine
```

**Comment by p12tic** [comment]
I think this bug has been fixed in #5599.
